### PR TITLE
RavenDB-19149 Optimize: Scope for Task && Cancellation Token Support

### PR DESCRIPTION
**Pull request by maciejaszyk**
https://issues.hibernatingrhinos.com/issue/RavenDB-19149/Check-if-Optimize-works-in-Compaction

### CODE REVIEW

File: src/Lucene.Net/Index/IndexWriter.cs [hunk 1]
```diff
@@ -2669,11 +2687,17 @@ internal System.String NewSegmentName()
 		/// </summary>
 		/// <throws>  CorruptIndexException if the index is corrupt </throws>
 		/// <throws>  IOException if there is a low-level IO error </throws>
-		/// <seealso cref="Index.LogMergePolicy.FindMergesForOptimize">
+		/// <seealso cref="Net.Index.LogMergePolicy.FindMergesForOptimize">
 		/// </seealso>
-		public virtual void  Optimize(IState state)
+		public virtual void  Optimize(IState state, CancellationToken token)
```
Comment (arekpalinski) [inline src/Lucene.Net/Index/IndexWriter.cs]: We can assign default value to `token` so you won't need to add `CancellationToken.None` in all places

File: src/Lucene.Net/Index/SegmentMerger.cs [hunk 2]
```diff
@@ -44,6 +44,38 @@ namespace Lucene.Net.Index
...
+			/// have been done since this method was last called.
+			/// When adding time-consuming code into SegmentMerger,
+			/// you should test different values for units to ensure
+			/// that the time in between calls to merge.checkAborted
+			/// is up to ~ 1 second.
+			/// </summary>
+			public override void Work(double units, IState state)
+			{
+				workCount += units;
+				if (workCount >= 10000.0)
```
Comment (arekpalinski) [inline src/Lucene.Net/Index/SegmentMerger.cs:66]: It is the same logic as in the base class, right?

File: src/Lucene.Net/Index/IndexWriter.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Lucene.Net/Index/IndexWriter.cs]: `CancellationToken token = default(CancellationToken)`

Comment (maciejaszyk) [reply src/Lucene.Net/Index/IndexWriter.cs]: done

File: src/Lucene.Net/Index/SegmentMerger.cs (hunk 2 above)
Comment (maciejaszyk) [reply src/Lucene.Net/Index/SegmentMerger.cs:66]: It has an additional check related to CancellationToken.

**Pull request merged**